### ISSUE TITLE
Saving a document over 1MB is blocking pending write queue

### ISSUE THREAD

**Issue by berkayulke** [firebase/firebase-js-sdk#4958]
### [REQUIRED] Describe your environment

  * Operating System version: PopOS 20.10
  * Browser version: Brave Version 1.25.68 Chromium: 91.0.4472.77 (Official Build) (64-bit)
  * Firebase SDK version: 8.2.4.
  * Firebase Product: firestore

## What happened?
We have an Angular v11.2.9 app with AngularFire v6.1.5 with persistence enabled. Some of our users reported that their changes were not showing up on other's computers. A few days later same users also reported that their save operations were never finishing.

After some debugging we discovered that at one point our app was trying to save an object with tons of small properties that add up to 1.5MB size without our error handling catching it and after causing that error, none of the save operations reached other users while they were seeing those changes on their app. And since those users were getting updates from other users they realized that bug so late that until we fixed that bug they had to keep asking each other if their changes were saved or not.

## How to reproduce?
I reproduced the same error, prepared a demo and deployed it with our setup on production. You can reach it on https://pending-write-error.web.app/

The code is here https://github.com/berkayulke/Pending-Write-Error-Demo

In there, try to save a huge document a few times, try to save a small one afterwards and check how it changes valueChanges and filtered snapshotChanges observables.

I added ongoing operation counter that show increases when you start a save and decreases when it is done or it fails.

Demo also calls waitForPendingWrites and timeouts after a few seconds for checking whether or not app has pending writes or not.

Long story short, try to save a document with lots of small properties that makes a huge object and it will block pending writes queue.

**Comment by looptheloop88** [comment]
Hi @berkayulke, thanks for the report. It looks like you’ve really put some thought and energy into this problem. Thanks for creating the sample project, we appreciate it.

First off, please confirm if I'm using the demo right based on the steps I did below:

1. Tried to save a small document and it was successful.
2. Tried to save a large document. I was able to see a value in valueChanges, however the data isn't successfully saved. After around 30 seconds of waiting, I was able to see this error in the browser console logs: `FirebaseError: A document cannot be written because it exceeds the maximum size allowed.` I'm not really sure if this delay came from generating the test object or the promise isn't return right away.

Based on your issue description, I shouldn't be able to save data again unless I clear the browser cache. Is this correct? This is not what I'm seeing in the demo. I was able to save small document again after the error was thrown.

**Comment by berkayulke** [comment]
Hi @looptheloop88, thanks for your return and using the demo.

Yes, you were using the demo as I described and I expected you to not be able to save data without clearing your cache.

As I was retesting with your steps and I realized that after around 1:20 minutes the promise was getting rejected and the small document's operation was resolved successfully, just as you described. I have a poor connection and my upload speed was 1 Mbps when I was testing so I assume this is the reason for the difference between our times.

Sorry for not waiting for it to finish before, I also want to tell you that I talked with my colleges and in our case, we were saving 2 documents over 1 mb at the same time. Also, the bugged feature is a frequently used one. We expect users to use that feature multiple times each session.

Today I tested the with the same demo and I think I found something that might be causing the error in our case.

The bugged feature was about our notifications, so it makes sense that first thing a user do after logging in is checking their notifications. While testing that case I did the following and produce an error and after the error the promise did not resolve or reject:

1. I first cleared my browser's cache by right clicking the refresh button on brave browser and clicking clear cache and hard reload and I was in incognito.
2. I tried to write few huge documents and before they resolve I refreshed the app.
3. I tried to save a single huge document.

After some time I saw the following errors on the console:

```
@firebase/firestore: Firestore (8.6.3): FirebaseError: [code=resource-exhausted]: Write stream exhausted maximum allowed queued writes
@firebase/firestore: Firestore (8.6.3): Using maximum backoff delay to prevent overloading the backend.
```

To make sure that I got it right this time I tried that multiple times and it was consistently doing the same thing. I even got a screencast in case you can't reproduce yourself. Here is the [link](https://drive.google.com/file/d/1iXHr6cBTBFHFN2x2r3gYBfGRQkUcK7Ze/view?usp=sharing). In the cast I first try to save a huge document and wait for it to reject, but I was able to reproduce without that so it's irrelevant. 

I left another tab open for 10 minutes and it was just the same 2 errors on my console. Here is a [screenshot](https://drive.google.com/file/d/1brik0VVDyX4EYwwIYFt8zZzVktaTredy/view?usp=sharing).

**Comment by looptheloop88** [comment]
Hi @berkayulke, thanks for the confirmation. Still, I wasn't able to reproduce the issue. I tried to write a large document multiple times, but the callback successfully fired after some time. There's a delay as you've mentioned, maybe due to internet connectivity speed.

The `resource-exhausted` error means that you exceeded one of the limits as described in the [documentation](https://firebase.google.com/docs/firestore/quotas). You may also want to check the Usage tab of your Firestore Database in Firebase console.

**Comment by schmidt-sebastian** [comment]
I also see the expected error in my repro:

```
[2021-06-04T16:16:35.952Z]  @firebase/firestore: Firestore (8.6.2): Connection GRPC stream error. Code: 3 Message: 3 INVALID_ARGUMENT: A document cannot be written because it exceeds the maximum size allowed.
Error [FirebaseError]: 3 INVALID_ARGUMENT: A document cannot be written because it exceeds the maximum size allowed.
```

This causes the SDK to reject the write, but other processing continues. Are you sure your write queue is blocked and not just slow?
